pyzeebe/grpc_internals/zeebe_adapter_base.py

```diff
@@ -46,34 +49,57 @@ def _create_channel(connection_uri: str, credentials: BaseCredentials = None,
 
     def _check_connectivity(self, value: grpc.ChannelConnectivity) -> None:
         logger.debug(f"Grpc channel connectivity changed to: {value}")
+        self.connected = False
+
         if value in [grpc.ChannelConnectivity.READY, grpc.ChannelConnectivity.IDLE]:
             logger.debug(f"Connected to {self.connection_uri or 'zeebe'}")
             self.connected = True
             self.retrying_connection = False
+            self._current_connection_retries = 0
+
         elif value == grpc.ChannelConnectivity.CONNECTING:
             logger.debug(f"Connecting to {self.connection_uri or 'zeebe'}.")
-            self.connected = False
             self.retrying_connection = True
+
         elif value == grpc.ChannelConnectivity.TRANSIENT_FAILURE:
-            logger.warning(f"Lost connection to {self.connection_uri or 'zeebe'}. Retrying...")
-            self.connected = False
-            self.retrying_connection = True
+            if self._should_retry():
+                logger.warning(f"Lost connection to {self.connection_uri or 'zeebe'}. Retrying...")
+                self.retrying_connection = True
+                self._current_connection_retries = self._current_connection_retries + 1
+            else:
+                logger.error(f"Failed to establish connection to {self.connection_uri or 'zeebe'}. Not recoverable")
+                self._close()
+                self.retrying_connection = False
+                raise ConnectionAbortedError(f"Lost connection to {self.connection_uri or 'zeebe'}")
+
         elif value == grpc.ChannelConnectivity.SHUTDOWN:
-            logger.error(f"Failed to establish connection to {self.connection_uri or 'zeebe'}. Non recoverable")
-            self.connected = False
+            logger.warning(f"Shutting down grpc channel to {self.connection_uri or 'zeebe'}")
             self.retrying_connection = False
-            raise ConnectionAbortedError(f"Lost connection to {self.connection_uri or 'zeebe'}")
+
+    def _should_retry(self):
+        return self._max_connection_retries == -1 or self._current_connection_retries < self._max_connection_retries
 
     def _common_zeebe_grpc_errors(self, rpc_error: grpc.RpcError):
         if self.is_error_status(rpc_error, grpc.StatusCode.RESOURCE_EXHAUSTED):
             raise ZeebeBackPressure()
         elif self.is_error_status(rpc_error, grpc.StatusCode.UNAVAILABLE):
+            self._current_connection_retries += 1
+            if not self._should_retry():
+                self._close()
             raise ZeebeGatewayUnavailable()
         elif self.is_error_status(rpc_error, grpc.StatusCode.INTERNAL):
+            if not self._should_retry():
+                self._close()
```

Comment:
This is lacking tests (ref. https://coveralls.io/builds/36254772/source?filename=pyzeebe/grpc_internals/zeebe_adapter_base.py). Tested manually, and it's working as expected.

Reply:
Added a test for this.
This led me to discover that `_current_connection_retries` was not incremented, so good catch!